### ISSUE TITLE
Rule Request: Warn on use of some NS types in Swift code

### ISSUE THREAD

**Issue by 72A12F4E** [realm/SwiftLint#2758]
### New Issue Checklist

- [X] Updated SwiftLint to the latest version
- [X] I searched for [existing GitHub issues](https://github.com/realm/SwiftLint/issues)

### New rule request

Please describe the rule idea, format
this issue's title as `Rule Request: [Rule Name]` and describe:

1. Why should this rule be added? Share links to existing discussion about what
   the community thinks about this.

This would prevent developers from using legacy NS types where swift versions of those types already exist.

2. Provide several examples of what _would_ and _wouldn't_ trigger violations.

**Bad**

All of the below types have a Swift analogue and should not be used in Swift code

```swift
let foo1: NSDictionary = bar1
let foo2 = NSArray()
let foo3: NSString = bar3
let foo4 = NSSet()
```

**Ok**

These types do not have a Swift analogue, so they should be allowed.

```swift
let vc = NSViewController()
let orderedSet = NSOrderedSet()
```

3. Should the rule be configurable, if so what parameters should be configurable?

Unsure

4. Should the rule be opt-in or enabled by default? Why?
   See [README.md](../README.md#opt-in-rules) for guidelines on when to mark a rule as opt-in.

I think this rule should be opt-in by default. There is a lot of potential for flagging types that should be completely valid in normal Swift code, and in codebases where very particular Swift-ObjC is required.

**Comment by 72A12F4E** [comment]
Here is a regex based rule that I implemented to identify these issues

```iml
custom_rules:
  legacy_objc_type:
    included: ".*.swift"
    name: "Legacy Obj-C Type"
    regex: "NSString|NSSet|NSArray|NSDictionary"
    message: "Avoid using legacy ObjC types where Swift types will suffice."
    severity: warning
    match_kinds:
      - typeidentifier
```
